### PR TITLE
Container image build

### PR DESCRIPTION
**Pull request by mwitzenm**
#### What is this PR About?
New guide for advanced image build
#### How should we test or review this PR?
pull it down and ensure content is correct
#### Is there a relevant Trello card or Github issue open for this?
https://trello.com/c/FW6WATci/51-apponb-5-and-apponb-6-write-an-image-building-best-practices-guide-and-dependency-analysis-for-dockerfile-creation
#### Who would you like to review this?
cc: @redhat-cop/cant-contain-this

### CODE REVIEW

File: playbooks/fundamentals/Container_Image_Build_Guidelines.adoc [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+
+
+We normally start with an existing image that is stored in a registry like the one provided by Red Hat http://registry.access.redhat.com/[here].  The images stored here were also started from a “base” image.  Each image instantiated by adding to a new formed Dockerfile is another layer(s) added to the starting base image.  Let’s begin to build our very first image by starting with an example Dockerfile and get into the details as we go along.  Edit our file with an IDE or plain text editor like below:
+
+
+vim Dockerfile
+
+----
```
Comment (etsauer) [inline playbooks/fundamentals/Container_Image_Build_Guidelines.adoc]: include `[source]`

File: playbooks/fundamentals/Container_Image_Build_Guidelines.adoc [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+The *MAINTAINER* line helps track changes to this new “base” image
+
+
+
+Now that we have these first two steps inside our file, we could run this container with a simple command of: docker build --rm -t rhel . (notice the . for the current working directory where our Dockerfile exists).  This image does not really do anything so let’s make it do something interesting like run a web server:
+
+
+
+
+----
```
Comment (etsauer) [inline playbooks/fundamentals/Container_Image_Build_Guidelines.adoc]: `[source]`

File: playbooks/fundamentals/Container_Image_Build_Guidelines.adoc [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+The *ENTRYPOINT* line is used to signify the command that will be run as if we were on the command line and this is the main execution process.
+
+The *CMD* line should be used to run the software contained by our image, along with any arguments. 
+
+
+
+== Onto the advanced use case and best practices
+
+==== Size does matter
```
Comment (etsauer) [inline playbooks/fundamentals/Container_Image_Build_Guidelines.adoc]: Don't skip header levels here. see the [style guide](https://github.com/redhat-cop/openshift-playbooks/blob/master/style_guide.adoc)

File: playbooks/fundamentals/Container_Image_Build_Guidelines.adoc [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+RUN yum -y install httpd && yum -y update; yum clean all
+
+----
+Another common procedure to ensure a layer is compact is to run a script referenced outside of the file to accomplish all our desired tasks.
+
+
+
+
+
+==== Conclusion
```
Comment (etsauer) [inline playbooks/fundamentals/Container_Image_Build_Guidelines.adoc]: Conclusion shouldn't be this deep in the structure.

File: playbooks/fundamentals/Container_Image_Build_Guidelines.adoc [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+* Trusted source of the starting point image -> The ability to trust the source of our images with backed proof of CVE’s is paramount to our security. 
+* Keep our image stateless ( attach a volume when state is required ) -> Its ephemeral nature allows for rapid deployment and creation.
+* Run one process per container -> In this case less is more, one process per container embraces our desired end goal of a microservices architecture.
+* Design for reusability and maintainability -> Let's not reinvent the wheel by creating images that already exist and secondly share and evangelize your images.
+* Reduce the number of components to execute within our configuration file -> Consider using scripts that are called within the image to reduce the layers created.
+* Clean up our image with unwanted cache and build artifacts -> I clean image is just good practice, removing unneeded libraries reduce the size, increase usability, reduce the possible attack surface. 
+
+
+
+==== Appendix
```
Comment (etsauer) [inline playbooks/fundamentals/Container_Image_Build_Guidelines.adoc]: Appendix shouldn't be a level 4 header

File: playbooks/fundamentals/Container_Image_Build_Guidelines.adoc [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+=== Base image
+
+Once we have landed on our image style it’s time to consider additional items such as where we actually obtain the image from.  It is paramount that we trust the source that provides the base image to us.  Images from Red Hat are a trusted source and maintained with constant improvement and patches.  
+
+=== Security
+
+Our third consideration to our image standard is security.  Security has different areas of concern when implementing container images.  We mentioned the trusted source as one such aspect as well as patching for vulnerabilities, but we also have to consider who is ultimately going to run our container.  When installing packages to the base image we will need to be root to accomplish such tasks, but once those tasks that require root privilege are complete, we should change the running container user with a USER 1001 or similar line in our file before we run the container.
```
Comment (etsauer) [inline playbooks/fundamentals/Container_Image_Build_Guidelines.adoc]: the ability to build images that operate properly without privileges is maybe the most important skill we can transfer. some more discussion of that topic and some examples would be very impactful

File: playbooks/fundamentals/Container_Image_Build_Guidelines.adoc [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+=== Base image
+
+Once we have landed on our image style it’s time to consider additional items such as where we actually obtain the image from.  It is paramount that we trust the source that provides the base image to us.  Images from Red Hat are a trusted source and maintained with constant improvement and patches.  
+
+=== Security
+
+Our third consideration to our image standard is security.  Security has different areas of concern when implementing container images.  We mentioned the trusted source as one such aspect as well as patching for vulnerabilities, but we also have to consider who is ultimately going to run our container.  When installing packages to the base image we will need to be root to accomplish such tasks, but once those tasks that require root privilege are complete, we should change the running container user with a USER 1001 or similar line in our file before we run the container.  
+
+=== Storage
```
Comment (etsauer) [inline playbooks/fundamentals/Container_Image_Build_Guidelines.adoc]: the more I read this, the more I think the following two sections aren't relevant to a discussion of "image building best practices", but more generic to containers

File: playbooks/fundamentals/Container_Image_Build_Guidelines.adoc [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+=== Process of the container 
+
+As we have touched on speed, that brings us to the next consideration: don’t run more than one process in our single container unless absolutely needed.  This goes back to the small is fast idea and it will reduce the complexity of our container as well.  Reusability and maintainability are very important to our efficiency as well as the efficiency of others who can use our image for their needs.  
+
+=== Reduced entries
+
+Each line in the Dockerfile adds a layer to the image.  As mentioned above in our RUN command for installing packages, we executed a line continuation with one of the following, a backslash ( / ) for multiline read ability or a semicolon ( ; ) or double ampersands ( && ) to ensure all those RUN segments commands are bundled into one layer.  These concatenated RUN commands should be used when the intent is bound around a common goal such as the installation of packages and the cleaning out of the resulting cache required for the package install.  As we see in the example:
```
Comment (pabrahamsson) [inline playbooks/fundamentals/Container_Image_Build_Guidelines.adoc]: You still have a forward slash here `( / )`, should be `( \ )`

**Comment by pabrahamsson** [comment]
@mwitzenm my apologies, I'd forgotten about the additional steps required for running `httpd` as a non-privileged user (port > 1023, log files, pid files etc). In the context of this guide I think you can remove the `USER` statement (and explanation).
Again, I'm sorry for the inconvenience.

The rest of the changes look good, just need to change the `/` to `\` (see inline comment)

**Pull request merged**